### DISCUSSION TITLE
Pure pursuit does not work with generated path file

### DISCUSSION THREAD

**Discussion by Williaml21** [LemLib/LemLib#153] [Q&A] [answered]
I'm using the latest LemLib V0.5.0 Release Candidate 7 and developing autonomous with path generator.

I have an experimental path file, named "A1path2.txt" ready. As I understand, since V0.5.0 release, the path file lives in the "static" folder and will be compiled and built into the final image to upload to the brain. I followed the example and below is the asset code:

`// get a path used for pure pursuit
// this needs to be put outside a function
//ASSET(example_txt); // '.' replaced with "_" to make c++ happy
ASSET(A1path2_txt); // '.' replaced with "_" to make c++ happy`

After compilation and build, I can see "A1path2.txt.o" being generated in "/bin/static" folder. And below is the code to use this path. 

`chassis.follow(A1path2_txt, 15, 100000, true, true);`

But when I run the code on robot, it doesn't move at all. What am I missing here?

Thanks,
William

**Comment by Raptorly1** [comment]
Hi William! If you could upload your path and maybe more of your code, that would be helpful. Double-check if you set the chassis pose close enough to the path. 

You can also consider joining the discord for faster/more responses https://discord.gg/3qEqmX6pXz

**Reply by Williaml21** [reply]
Please find attached path file.

There isn't much code for this test other than the ASSET path and calling follow function.
Now that you mentioned chassis pose being set close enough, I did reset the chassis pose to (0,0) at initialization. I can see the pose data showing all zeros on the brain LCD at the start. Could that be a problem?

[A1path2.txt](https://github.com/LemLib/LemLib/files/14784742/A1path2.txt)

`/**
 * runs after initialize if the robot is connected to field control
 */
void competition_initialize() {
    resetDriveEncoders();

    autonomous();
}

// get a path used for pure pursuit
// this needs to be put outside a function
//ASSET(example_txt); // '.' replaced with "_" to make c++ happy
ASSET(A1path2_txt); // '.' replaced with "_" to make c++ happy

/* Helper functions for autonomous*/
    

/**
 * Runs during auto
 *
 * This is an example autonomous routine which demonstrates a lot of the features LemLib has to offer
 */
void autonomous() {
    pros::lcd::print(0, "Show me your coordinates:"); // x

    chassis.follow(A1path2_txt, 15, 100000, true, true);

}`

**Reply by Williaml21** [reply]
I wanted the robot starting point to be (0,0), then the path should be referenced to that origin. But I realized that the path generator has the center of field as origin. How should I generate the working path file? Offset the path data by robot's starting point, say (-36, -58)?

**Reply by Williaml21** [reply]
BTW, when I clicked on the discord link in your message, it shows "invalid invite" and I couldn't join. Why is that?

**Reply by Raptorly1** [reply]
Your robot starts at (0,0) which is too far away from your path which starts at approximately (-35.859,-58.103).

Side note: your paths should not overlap. Instead, make multiple path files.

Let me know if this helps!

**Reply by Raptorly1** [reply]
you can do `chassis.setPose(x,y,theta)`

**Reply by Raptorly1** [reply]
I'm not sure why the discord link doesn't work. Maybe try this one https://discord.gg/mG7t9TvN3q

**Reply by Williaml21** [reply]
Thanks, I'll give it a try as you suggested.

However, when I run the example path file, it doesn't move either. The example path does start from (0,0,100) though.

**Reply by Raptorly1** [reply]
Does the brain display any errors?

**Reply by Williaml21** [reply]
no error.

Update: I just set the starting pose close to my path' starting position, and it does work now. Although, the actual route is a bit far from my path drawing. It seems needing further fine tune. But at least the basic function seems working now.

**Reply by Williaml21** [reply]
I'm skeptical about my IMU sensor. Even though it's connected to the port, but not entirely sure if it's really functioning as expected. Is there any quick way to verify its functionality and accuracy?

**Reply by Raptorly1** [reply]
Great to hear!

**Reply by Raptorly1** [reply]
You can try moving the robot around manually and seeing if the theta displayed is accurate

**Reply by Williaml21** [reply]
That's exactly why I'm skeptical about IMU. I did manually rotate the robot, and didn't see theta change much. I'm new to PROS, not sure how to debug live where I could print variables out to the terminal. Any suggestions?

**Reply by Williaml21** [reply]
But when I disconnect the IMU cable from the port, the brain does warn losing Inertial connection. So it seems being detected. But just not functions as expected. And I did reset and calibrate IMU at the init. not sure why not working.

**Reply by Williaml21** [reply]
I add a line to print IMU.ratation() data on brain LCD, it shows inf. So it means it's working. Could it fail calibration?

`pros::lcd::print(4, "IMU Rotation: %f", imu.get_rotation()); // IMU rotation`

**Reply by Rocky14683** [reply, marked as answer]
yes that could be a calibration issue, the issue have been fix in PROS kernel 3.8.3. So the next release of lemlib will allow the imu to calibrate.

**Reply by Raptorly1** [reply]
@Williaml21 You can try using lemlib 0.5.0-rc.6 and PROS kernel@3.8.0. That version is confirmed to not have any IMU calibration issues
